### PR TITLE
CryoSECOM position switching missing functions 

### PR DESCRIPTION
**Pull request by lazem**
Some further missing features in the `move` module for the CryoSECOM position switching procedures: 

- Add support for safe movement to coating position
- Add function returning current position label 
- Rename `getLoadingProgress` to `getMovementProgress`
- Refine `move_test` test cases

### CODE REVIEW

File: src/odemis/acq/move.py [hunk 1]
```diff
@@ -35,19 +34,68 @@
 from odemis.util import executeAsyncTask
 
 MAX_SUBMOVE_DURATION = 60  # s
-LOADING, IMAGING = 0, 1
+LOADING, IMAGING, TILTED, COATING, IMAGING_PATH, COATING_PATH, UNKNOWN = 0, 1, 2, 3, 4, 5, 6
```
Comment (lazem) [inline src/odemis/acq/move.py:37]: Since both movements to imaging and coating positions starts from the loading position, renamed `LOADING_PATH` to `IMAGING_PATH` to better reflect the destination.

File: install/linux/usr/share/odemis/sim/cryosecom-sim.yaml [hunk 2]
```diff
@@ -164,7 +164,7 @@ CryoSECOM: {
   class: tmcm.TMCLController,
   init: {
     port: "/dev/fake6",
-    ustepsize: [5.9e-9, 5.9e-9, 5.9e-9, 5.9e-9, 5.9e-9],
+    ustepsize: [5.9e-9, 5.9e-9, 5.9e-9, 5.9e-6, 5.9e-6],
```
Comment (lazem) [inline install/linux/usr/share/odemis/sim/cryosecom-sim.yaml:167]: Had to lower ustepsize for `rx` and `rx`, otherwise simulator times out before getting to coating `rx` angel = 0.4363

File: src/odemis/acq/move.py [hunk 3]
```diff
@@ -77,32 +125,30 @@ def check_axes(pos):
         return None
 
 
-def _isInRange(stage, axis):
+def _isInRange(current_pos, active_range, axis):
```
Comment (Kleijwegt) [inline src/odemis/acq/move.py:128]: I think that for a couple of functions you forgot to add some input values to the docstring

File: src/odemis/acq/move.py [hunk 4]
```diff
@@ -77,32 +125,30 @@ def check_axes(pos):
...
+    margin = (axis_active_range[1] - axis_active_range[0]) * 0.01
+    return (axis_active_range[0] - margin) <= pos <= (axis_active_range[1] + margin)
 
 
-def _isNearDeactive(current_pos, stage_deactive, axis):
+def _isNearPosition(current_pos, target_position, axis):
     """
-    Check whether given axis is near stage's FAV_POS_DEACTIVE position
-    :return: True if the axis is near (within 1 mm or 0.1 rad) FAV_POS_DEACTIVE, False otherwise
+    Check whether given axis is near stage target position
```
Comment (Kleijwegt) [inline src/odemis/acq/move.py:142]: docustring does not hold all the input values

File: src/odemis/acq/test/move_test.py [hunk 5]
```diff
@@ -71,75 +73,101 @@ def setUp(self):
...
+        f = cryoLoadSample(COATING)
+        f.result()
+        with self.assertRaises(ValueError):
+            f = cryoLoadSample(IMAGING)
+            f.result()
+
+        # Test tilting from coating
         with self.assertRaises(ValueError):
-            f = cryoTiltSample(rx=0.003, rz=0.003)
+            f = cryoTiltSample(rx=self.tilt_angle, rz=self.tilt_angle)
```
Comment (pieleric) [inline src/odemis/acq/test/move_test.py:170]: It doesn't really make sense to use the same angle for both, as they are unrelated. So at least, use different variable names.
Also, please test with rz=0, it's a corner-case which is going to be very common.

**Review by pieleric — CHANGES_REQUESTED** [review]
Looks fine. I mostly have comments to simplify the code, and improve the test cases.

File: src/odemis/acq/move.py [hunk 6]
```diff
@@ -34,10 +34,10 @@
 from odemis.util import executeAsyncTask
 
 MAX_SUBMOVE_DURATION = 60  # s
-LOADING, IMAGING, TILTED, COATING, IMAGING_PATH, COATING_PATH, UNKNOWN = 0, 1, 2, 3, 4, 5, 6
+LOADING, IMAGING, TILTED, COATING, LOADING_PATH, UNKNOWN = 0, 1, 2, 3, 4, 5
 ATOL_LINEAR_POS = 100e-6  # m
 ATOL_ROTATION_POS = 1e-3  # rad (~0.5°)
-RTOL_PROGRESS = 0.1
+RTOL_PROGRESS = 0.3
```
Comment (lazem) [inline src/odemis/acq/move.py:40]: Increased this value to reflect the results I got in the real stage, but may need more testing.

File: src/odemis/acq/move.py [hunk 7]
```diff
@@ -34,10 +34,10 @@
 from odemis.util import executeAsyncTask
 
 MAX_SUBMOVE_DURATION = 60  # s
-LOADING, IMAGING, TILTED, COATING, IMAGING_PATH, COATING_PATH, UNKNOWN = 0, 1, 2, 3, 4, 5, 6
+LOADING, IMAGING, TILTED, COATING, LOADING_PATH, UNKNOWN = 0, 1, 2, 3, 4, 5
```
Comment (lazem) [inline src/odemis/acq/move.py:37]: As the coating position is now in the imaging range, returned it back to only one path `LOADING_PATH`

File: src/odemis/acq/move.py [hunk 8]
```diff
@@ -197,22 +206,12 @@ def _doCryoLoadSample(future, stage, focus, target):
...
-                for axis in {'x', 'y', 'z'}:
-                    if not _isInRange(current_pos, stage_active_range, axis):
-                        logging.warning("Axis %s position is out of active range.", axis)
-                # Check that current position is near the ACTIVE to DEACTIVE line
-                # (ie, in case it was previously stopped half-way through loading/unloading)
-                if getMovementProgress(stage.position.value, stage_active, stage_deactive) is None:
-                    logging.warning("Stage position is not near the ACTIVE to DEACTIVE line.")
-
+            if getCurrentPositionLabel(current_pos, stage) is UNKNOWN:
+                logging.warning("Current stage position is unknown.")
```
Comment (lazem) [inline src/odemis/acq/move.py:214]: That's enough, I believe!

File: src/odemis/acq/move.py [hunk 9]
```diff
@@ -234,24 +289,29 @@ def _doCryoTiltSample(future, stage, focus, rx, rz):
...
-                util.almost_equal(stage.position.value['rz'], 0, atol=ATOL_ROTATION_POS)):
-            sub_moves.append((focus, focus.getMetadata()[model.MD_FAV_POS_DEACTIVE]))
-        elif not util.almost_equal(focus.position.value, focus.getMetadata()[model.MD_FAV_POS_DEACTIVE],
-                                   atol=ATOL_LINEAR_POS):
-            raise ValueError("Cannot proceed while focus is not near FAV_POS_DEACTIVE position.")
+        # Park focus only if stage rx and rz are equal to 0
+        # Otherwise stop if it's not already parked
+        if not util.almost_equal(focus.position.value['z'], focus_deactive['z'], atol=ATOL_LINEAR_POS):
+            if (util.almost_equal(current_pos['rx'], 0, atol=ATOL_ROTATION_POS) and
+                    util.almost_equal(current_pos['rz'], 0, atol=ATOL_ROTATION_POS)):
```
[the remark this answers is not in the document]
Comment (lazem) [reply src/odemis/acq/move.py]: Nice suggestion, thanks! Extended `_isNearPosition` and used it for these checks.

**Comment by lazem** [comment]
Hi @pieleric Added the suggested changes plus the ones I found in testing. 
Appreciating your review. Thanks!

**Review by pieleric — CHANGES_REQUESTED** [review]
I only have a couple of small tweaks to recommend.

There are more high-level things to consider, mainly about the inter-position paths, and the rz importance. For now I'd suggest to keep them as it is, and afterwards reconsider adjust them... once we have a better idea.
